tools/clang/unittests/HLSLExec/LongVectors.cpp

```diff
@@ -2,25 +2,235 @@
 #include "HlslExecTestUtils.h"
 #include <iomanip>
 
-LongVector::BinaryOpType
+template <typename LongVectorOpTypeT>
+const LongVector::OpTypeMetaData<LongVectorOpTypeT> &
+LongVector::getLongVectorOpType(const OpTypeMetaData<LongVectorOpTypeT> *Values,
+                                const std::wstring &OpTypeString,
+                                std::size_t Length) {
+  for (size_t i = 0; i < Length; i++) {
+    if (Values[i].OpTypeString == OpTypeString)
+      return Values[i];
+  }
+
+  LOG_ERROR_FMT_THROW(L"Invalid LongVectorOpType string: %s",
+                      OpTypeString.c_str());
+
+  throw std::runtime_error("Invalid LongVectorOpType string");
+}
+
+const LongVector::OpTypeMetaData<LongVector::BinaryOpType> &
 LongVector::getBinaryOpType(const std::wstring &OpTypeString) {
   return getLongVectorOpType<LongVector::BinaryOpType>(
-      binaryOpTypeStringToEnumMap, OpTypeString,
-      std::size(binaryOpTypeStringToEnumMap));
+      binaryOpTypeStringToOpMetaData, OpTypeString,
+      std::size(binaryOpTypeStringToOpMetaData));
 }
 
-LongVector::UnaryOpType
+const LongVector::OpTypeMetaData<LongVector::UnaryOpType> &
 LongVector::getUnaryOpType(const std::wstring &OpTypeString) {
   return getLongVectorOpType<LongVector::UnaryOpType>(
-      unaryOpTypeStringToEnumMap, OpTypeString,
-      std::size(unaryOpTypeStringToEnumMap));
+      unaryOpTypeStringToOpMetaData, OpTypeString,
+      std::size(unaryOpTypeStringToOpMetaData));
 }
 
-LongVector::TrigonometricOpType
+const LongVector::OpTypeMetaData<LongVector::AsTypeOpType> &
+LongVector::getAsTypeOpType(const std::wstring &OpTypeString) {
+  return getLongVectorOpType<LongVector::AsTypeOpType>(
+      asTypeOpTypeStringToOpMetaData, OpTypeString,
+      std::size(asTypeOpTypeStringToOpMetaData));
+}
+
+const LongVector::OpTypeMetaData<LongVector::TrigonometricOpType> &
 LongVector::getTrigonometricOpType(const std::wstring &OpTypeString) {
   return getLongVectorOpType<LongVector::TrigonometricOpType>(
-      trigonometricOpTypeStringToEnumMap, OpTypeString,
-      std::size(trigonometricOpTypeStringToEnumMap));
+      trigonometricOpTypeStringToOpMetaData, OpTypeString,
+      std::size(trigonometricOpTypeStringToOpMetaData));
+}
+
+// Helper to fill the test data from the shader buffer based on type. Convenient
+// to be used when copying HLSL*_t types so we can use the underlying type.
+template <typename DataTypeT>
+void LongVector::fillLongVectorDataFromShaderBuffer(
+    MappedData &ShaderBuffer, std::vector<DataTypeT> &TestData,
+    size_t NumElements) {
+  if constexpr (std::is_same_v<DataTypeT, HLSLHalf_t>) {
```

Comment:
The `if {} else if {} else` here is a great candidate for early exiting

Reply:
Do we prefer that? Feels like it adds bloat given that there are no statements
outside of if blocks. I'm assuming yes and making the change anyways.

Reply:
https://llvm.org/docs/CodingStandards.html#use-early-exits-and-continue-to-simplify-code

Reply:
The section after the one I linked to makes an interesting point about how this may not do quite what you expect when using `if constexpr`.  I'm not sure if that applies in this case though.